src/community/label_propagation.c

```diff
@@ -26,15 +26,410 @@
+                    {
+                        /* Select randomly from the dominant labels */
+                        k = RNG_INTEGER(0, igraph_vector_int_size(&dominant_labels) - 1);
+                        k = VECTOR(dominant_labels)[(long int)k];
+                        /* If label changes, we will continue running */
+                        if (k != j)
+                            running = true;
+                        /* Actually change label */
+                        VECTOR(*membership)[v1] = k;
+                    }
+                } else {
+                    /* We are not using retention, so check if we should do a control iteration
+                    or an update iteration. */
+                    if (control_iteration) {
+                        /* Check if the _current_ label of the node is also dominant */
+                        k = VECTOR(*membership)[v1];
+                        if (k < 0 ||                              /* No label assigned yet or */
+                            VECTOR(label_counters)[k] < max_count /* Label is not maximum */
+                            ) {
+                            /* Nope, we need at least one more iteration */
+                            running = true;
+                        }
+                    }
+                    else {
+                        /* Select randomly from the dominant labels */
+                        k = RNG_INTEGER(0, igraph_vector_int_size(&dominant_labels) - 1);
+                        VECTOR(*membership)[v1] = VECTOR(dominant_labels)[k];
+                    }
+                }
+            }
+
+            /* Clear the nonzero elements in label_counters */
+            num_neis = igraph_vector_int_size(&nonzero_labels);
+            for (j = 0; j < num_neis; j++) {
+                VECTOR(label_counters)[VECTOR(nonzero_labels)[j]] = 0;
+            }
+        }
+
+        /* Alternating between control iterations and label updating iterations */
+        if (!retention)
+                control_iteration = !control_iteration;
+    }
+
+    RNG_END();
+
+    if (weights) {
+        igraph_inclist_destroy(&il);
+    } else {
+        igraph_adjlist_destroy(&al);
+    }
+    IGRAPH_FINALLY_CLEAN(1);
+
+    igraph_vector_int_destroy(&node_order);
+    igraph_vector_int_destroy(&nonzero_labels);
+    igraph_vector_int_destroy(&dominant_labels);
+    igraph_vector_destroy(&label_counters);
+    IGRAPH_FINALLY_CLEAN(4);
+
+    return IGRAPH_SUCCESS;
+}
+
+igraph_error_t igraph_i_community_fast_label_propagation(const igraph_t *graph,
+    igraph_vector_int_t *membership,
+    igraph_neimode_t mode,
+    const igraph_vector_t *weights,
+    igraph_vector_bool_t *fixed) {
+
+    igraph_integer_t no_of_nodes = igraph_vcount(graph), no_of_not_fixed_nodes = 0;
+    igraph_integer_t i, j, k;
+    igraph_inclist_t il;
+    igraph_adjlist_t al;
+
+    igraph_vector_t label_counters;
```

Comment:
There was a comment on this in the original code explaining why this is of real type, and not integer (despite the confusing name): `/* real type, stores weight sums */`. Please restore this comment, since it seems to be valid for this algorithm variant as well.

Reply:
Added.